# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieParser.cpp
# Commit: 675785853d7a
# Full Hash: 675785853d7a40123826e1b47c49d68f87bd4bd0
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-17 17:00:49
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r de6a26d3e3df -r 675785853d7a netwerk/cookie/CookieParser.cpp
--- a/netwerk/cookie/CookieParser.cpp	Tue Dec 17 08:57:31 2024 +0000
+++ b/netwerk/cookie/CookieParser.cpp	Tue Dec 17 08:57:31 2024 +0000
@@ -873,7 +873,7 @@
                          CookieStatus aStatus, nsCString& aCookieHeader,
                          const nsACString& aDateHeader, bool aFromHttp,
                          bool aIsForeignAndNotAddon, bool aPartitionedOnly,
-                         bool aIsInPrivateBrowsing) {
+                         bool aIsInPrivateBrowsing, bool aOn3pcbException) {
   MOZ_ASSERT(!mContainsCookie);
 
   // init expiryTime such that session cookies won't prematurely expire
@@ -1021,6 +1021,19 @@
     return newCookie;
   }
 
+  // If the cookie is on the 3pcd exception list, we apply partitioned
+  // attribute to the cookie.
+  if (aOn3pcbException) {
+    // We send a warning if the cookie doesn't have the partitioned attribute
+    // in the foreign context.
+    if (aPartitionedOnly && !mCookieData.isPartitioned() &&
+        aIsForeignAndNotAddon) {
+      mWarnings.mForeignNoPartitionedWarning = true;
+    }
+
+    mCookieData.isPartitioned() = true;
+  }
+
   // If the cookie does not have the partitioned attribute,
   // but is foreign we should give the developer a message.
   // If CHIPS isn't required yet, we will warn the console
@@ -1036,6 +1049,7 @@
       RejectCookie(RejectedForeignNoPartitionedError);
       return newCookie;
     }
+
     mWarnings.mForeignNoPartitionedWarning = true;
   }
 